src/main/java/duke/storage/Storage.java

```diff
@@ -0,0 +1,156 @@
+            while (fileReader.hasNextLine()) {
+                String entry = fileReader.nextLine();
+                try {
+                    ArrayList<String> parsedEntry = Parser.parseDatabaseEntry(entry);
+                    taskList.add(readEntry(parsedEntry));
+                } catch (DukeDatabaseInvalidEntryException e) {
+                    System.out.println(e);
+                } catch (DukeEndDateBeforeStartDateException e) {
+                    System.out.println(e);
+                }
+            }
+            fileReader.close();
+            return taskList;
+        } catch (FileNotFoundException e) {
+            throw new DukeDatabaseNotFoundException();
+        }
+    }
+
+    /**
+     * Saves the TaskList to the data file.
+     *
+     * @param taskList The TaskList containing the tasks.
+     */
+    public void saveData(TaskList taskList) {
+        try {
+            createDataDirectory();
+            createDukeFile();
+            writeTaskListToFile(taskList);
+        } catch (IOException e) {
+            System.out.println("Something went wrong: " + e.getMessage());
+        }
+    }
+
+    private Task readEntry(ArrayList<String> parsedEntry) throws DukeDatabaseInvalidEntryException,
+            DukeEndDateBeforeStartDateException {
+        Task newTask;
+        switch (parsedEntry.get(0)) {
+        case "T":
+            newTask = new Todo(parsedEntry.get(2));
+            if (parsedEntry.get(1).equals("X")) {
+                newTask.markAsDone();
+            }
+            break;
+        case "D":
+            newTask = new Deadline(parsedEntry.get(2), formatDate(parsedEntry.get(3)));
+            if (parsedEntry.get(1).equals("X")) {
+                newTask.markAsDone();
+            }
+            break;
+        case "E":
+            newTask = new Event(parsedEntry.get(2), formatDate(parsedEntry.get(3)),
+                    formatDate(parsedEntry.get(4)));
+            if (parsedEntry.get(1).equals("X")) {
+                newTask.markAsDone();
+            }
+            break;
+        default:
+            throw new DukeDatabaseInvalidEntryException();
+        }
+        return newTask;
+    }
+
+    private void createDataDirectory() throws IOException {
+        File dataDirectory = new File(this.filepath).getParentFile();
+        if (!dataDirectory.exists()) {
+            if (!dataDirectory.mkdirs()) {
+                throw new IOException("Failed to create data directory");
+            }
+        }
+    }
+
+    private void createDukeFile() throws IOException {
+        File dukeFile = new File(this.filepath);
+        if (!dukeFile.exists()) {
+            if (!dukeFile.createNewFile()) {
+                throw new IOException("Failed to create duke file");
+            }
+        }
+    }
+
+    private void writeTaskListToFile(TaskList taskList) throws IOException {
+        Path filePath = Paths.get(this.filepath);
+        try {
+            Files.deleteIfExists(filePath);
```

Comment:
is this unnecessary? when doing File file = new File(filePath); i believe it should already override the existing file in the filepath

Reply:
My idea is to overwrite the existing file rather than appending to an existing file. Please let me know if you have any alternative implementations !